runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/PipelineTranslator.java

```diff
@@ -17,170 +17,336 @@
-  //  Pipeline utility methods
-  // --------------------------------------------------------------------------------------------
+    @Override
+    public SerializablePipelineOptions getSerializableOptions() {
+      return serializableOptions;
+    }
 
-  /** Get a {@link TransformTranslator} for the given {@link TransformHierarchy.Node}. */
-  protected abstract @Nullable <
-          InT extends PInput, OutT extends POutput, TransformT extends PTransform<InT, OutT>>
-      TransformTranslator<InT, OutT, TransformT> getTransformTranslator(
-          @Nullable TransformT transform);
-
-  /** Apply the given TransformTranslator to the given node. */
-  private <InT extends PInput, OutT extends POutput, TransformT extends PTransform<InT, OutT>>
-      void applyTransformTranslator(
-          TransformHierarchy.Node node,
-          TransformT transform,
-          TransformTranslator<InT, OutT, TransformT> transformTranslator) {
-    // create the applied PTransform on the translationContext
-    AppliedPTransform<InT, OutT, PTransform<InT, OutT>> appliedTransform =
-        (AppliedPTransform) node.toAppliedPTransform(getPipeline());
-    try {
-      transformTranslator.translate(transform, appliedTransform, translationContext);
-    } catch (IOException e) {
-      throw new RuntimeException(e);
+    @Override
+    public SparkSession getSparkSession() {
+      return sparkSession;
     }
   }
 
-  // --------------------------------------------------------------------------------------------
-  //  Pipeline visitor entry point
-  // --------------------------------------------------------------------------------------------
-
   /**
-   * Translates the pipeline by passing this class as a visitor.
+   * {@link PTransformVisitor} that analyses dependencies of supported {@link PTransform
+   * PTransforms} to help identify cache candidates.
    *
-   * @param pipeline The pipeline to be translated
+   * <p>The visitor may throw if a {@link PTransform} is observed that uses unsupported features.
    */
-  public void translate(Pipeline pipeline) {
-    LOG.debug("starting translation of the pipeline using {}", getClass().getName());
-    pipeline.traverseTopologically(this);
+  private class DependencyVisitor extends PTransformVisitor {
+    private final Map<PCollection<?>, TranslationResult<?>> results = new HashMap<>();
+
+    @Override
+    <InT extends PInput, OutT extends POutput> void visit(
+        Node node,
+        PTransform<InT, OutT> transform,
+        TransformTranslator<InT, OutT, PTransform<InT, OutT>> translator) {
+      for (PCollection<?> pOut : node.getOutputs().values()) {
+        results.put(pOut, new TranslationResult<>(pOut));
+        for (Map.Entry<TupleTag<?>, PCollection<?>> entry : node.getInputs().entrySet()) {
+          TranslationResult<?> input = checkStateNotNull(results.get(entry.getValue()));
+          input.dependentTransforms.add(transform);
+        }
+      }
+    }
```

Comment:
I don't understand this algo especially the nested loops because `pOut` is not referenced inside the inner loop.
Why don't you populate `results` based on outputs first and then do the second loop?

There might be something I don't get, if it's the case, please add some comments to the algo for ease of maintenance.

Reply:
you're right, this isn't necessary anymore, thx for catching 👍  in an earlier version i was tracking downstream dependencies on the pCollection level rather than the transform level, which isn't actually necessary.

Reply:
this should be 
```java
      // add new translation result for every output of `transform`
      for (PCollection<?> pOut : node.getOutputs().values()) {
        results.put(pOut, new TranslationResult<>(pOut));
      }
      // track `transform` as downstream dependency for every input
      for (Map.Entry<TupleTag<?>, PCollection<?>> entry : node.getInputs().entrySet()) {
        TranslationResult<?> input = checkStateNotNull(results.get(entry.getValue()));
        input.dependentTransforms.add(transform);
      }
    }
```

Reply:
yes, that is exactly the code I mentioned here
> Why don't you populate results based on outputs first and then do the second loop?

:+1: